### ISSUE TITLE
Version field?

### ISSUE THREAD

**Issue by Gunzter** [core-wg/oscoap#183]
I realized that there are no way of communicating the version used by a node using OSCORE, the only thing that will happen is that the decryption/authentication will fail. Maybe it would be a good idea to specify how to communicate what version a node is using to avoid potential problems down the road?
My spontaneous solution is to send the version number in the "Error Message" when decryption/authentication fails.

**Comment by vikram919** [comment]
Hi!
I think it should be more easier to access. Adding version number to the firstbyte of compressed cose message just like partialIV length?

**Comment by fpalombini** [comment]
that would add 1B to all messages...

**Comment by Gunzter** [comment]
Some more ideas I got:
Use the CoAP version field - This should probably not be touched.
Use one or more of the three unused bits in the compressed COSE header - They are probably reserved for Multicast?
Use a new CoAP option number for a new OSCORE version - I don't know how this works with IANA considerations, but can this be a possibility? Using a new option number forces the server to respond respond with 4.02 (Bad Option), this can then be handled.

**Comment by LudwigSeitz** [comment]
I like the last one best. Using new CoAP option numbers for new OSCORE versions seems like a very clean solution, that would be backwards compatible and not affect the overhead of the messages.

**Comment by cabo** [comment]
I was already going to ask what point a version number was going to have.  We have enough extension points, and the option number is the obvious place to do a catastrophic upgrade.  (Apart from that function, version numbers make little sense anyway.)

**Comment by gselander** [comment]
We previously integrity protected the CoAP version number to mitigate against some yet-unknown  attack mixing CoAP versions. In order to comply with end-to-end secure REST (e.g. HTTP client to CoAP server) we replaced the integrity protection of CoAP version with the integrity protection of OSCORE version. So, if there is a future version of CoAP that changes the CoAP messages in such a way that this version (1) of OSCORE does not provide adequate protection, then a new version of OSCORE can be defined and distinguished using the integrity of OSCORE version.

This construction does not allow the recipient to tell that the verification error was caused by a different OSCORE version, but as mentioned there are extension points to add that should it become necessary.
